### PR TITLE
Fix Promise.any Reject Element Function

### LINKED ISSUE #47: Spec bug - Promise.any(elements) should be rejected if all elements are rejected
[`Promise.any` Reject Element Functions](https://tc39.es/proposal-promise-any/#sec-promise.any-reject-element-functions) step 11 says:

> 0. If *remainingElementsCount*.[[Value]] is 0, then
>     0. Perform ? ThrowAggregateError(*errors*).

I think this rejects the wrong promise. It should behave like `Promise.all` and `Promise.any`, and deliver the results to the *promiseCapability* fetched in step 7. (Currently that capability is unused.)

### PR DESCRIPTION
**Pull request by chicoxyzzy**
Fixes https://github.com/tc39/proposal-promise-any/issues/47

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -264,7 +264,7 @@ <h1>`Promise.any` Reject Element Functions</h1>
       1. Set _errors_[_index_] to _x_.
       1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
       1. If _remainingElementsCount_.[[Value]] is 0, then
-        1. Perform ? ThrowAggregateError(_errors_).
+        1. Return ? Call(_promiseCapability_.[[Resolve]], undefined, &laquo; ThrowAggregateError(_errors_) &raquo;).
```
Comment (ljharb) [inline spec.html]: I think the abstract operation returns an abrupt completion, so you’d need to construct the error without throwing it?

**Review by ljharb — CHANGES_REQUESTED** [review]

File: spec.html (hunk 1 above)
Comment (chicoxyzzy) [reply spec.html]: You are right. In that case, I think, we don't need that abstract operation at all. I removed it in the next commit.

File: spec.html [hunk 2]
```diff
@@ -228,7 +228,9 @@ <h1>Runtime Semantics: PerformPromiseAny ( _iteratorRecord_, _constructor_, _res
           1. Set _iteratorRecord_.[[Done]] to *true*.
           1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
           1. If _remainingElementsCount_.[[Value]] is 0, then
-            1. Perform ? ThrowAggregateError(_errors_).
+            1. Let _error_ be a newly created `AggregateError` object.
+            1. Set _error_.[[AggregateErrors]] to _errors_.
+            1. Return ThrowCompletion ( _error_ ).
```
Comment (ljharb) [inline spec.html]:
```suggestion
            1. Return ThrowCompletion(_error_).
```

**Review by ljharb — APPROVED** [review]

File: spec.html [hunk 3]
```diff
@@ -228,7 +228,9 @@ <h1>Runtime Semantics: PerformPromiseAny ( _iteratorRecord_, _constructor_, _res
           1. Set _iteratorRecord_.[[Done]] to *true*.
           1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
           1. If _remainingElementsCount_.[[Value]] is 0, then
-            1. Perform ? ThrowAggregateError(_errors_).
+            1. Let _error_ be a newly created `AggregateError` object.
```
Comment (mathiasbynens) [inline spec.html]:
```suggestion
            1. Let _error_ be a newly-created `AggregateError` object.
```

File: spec.html [hunk 4]
```diff
@@ -264,24 +266,15 @@ <h1>`Promise.any` Reject Element Functions</h1>
       1. Set _errors_[_index_] to _x_.
       1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
       1. If _remainingElementsCount_.[[Value]] is 0, then
-        1. Perform ? ThrowAggregateError(_errors_).
+        1. Let _error_ be a newly created `AggregateError` object.
```
Comment (mathiasbynens) [inline spec.html]:
```suggestion
        1. Let _error_ be a newly-created `AggregateError` object.
```

**Review by mathiasbynens — CHANGES_REQUESTED** [review]

**Comment by chicoxyzzy** [comment]
@mathiasbynens @bakkot Did you get a chance to look into this?

File: spec.html [hunk 5]
```diff
@@ -228,7 +228,9 @@ <h1>Runtime Semantics: PerformPromiseAny ( _iteratorRecord_, _constructor_, _res
           1. Set _iteratorRecord_.[[Done]] to *true*.
           1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
           1. If _remainingElementsCount_.[[Value]] is 0, then
-            1. Perform ? ThrowAggregateError(_errors_).
+            1. Let _error_ be a newly-created `AggregateError` object.
```
Comment (bakkot) [inline spec.html:231]:
```suggestion
            1. Let _error_ be a newly created `AggregateError` object.
```

(Every other use of "newly created" in the spec is spelled this way; cc @mathiasbynens.)

Comment (mathiasbynens) [reply spec.html:231]: Sure, but that’s wrong.

File: spec.html [hunk 6]
```diff
@@ -264,24 +266,15 @@ <h1>`Promise.any` Reject Element Functions</h1>
       1. Set _errors_[_index_] to _x_.
       1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
       1. If _remainingElementsCount_.[[Value]] is 0, then
-        1. Perform ? ThrowAggregateError(_errors_).
+        1. Let _error_ be a newly-created `AggregateError` object.
+        1. Set _error_.[[AggregateErrors]] to _errors_.
+        1. Return ? Call(_promiseCapability_.[[Resolve]], undefined, &laquo; _error_ &raquo;).
```
Comment (bakkot) [inline spec.html]:
```suggestion
        1. Return ? Call(_promiseCapability_.[[Reject]], undefined, &laquo; _error_ &raquo;).
```

File: spec.html (hunk 5 above)
Comment (bakkot) [reply spec.html:231]: I think proposals should follow the style conventions in the upstream spec. If you want it to change here, it should change upstream first.

File: spec.html (hunk 6 above)
Comment (chicoxyzzy) [reply spec.html]: Thank you!

**Review by bakkot — APPROVED** [review]

File: spec.html (hunk 5 above)
Comment (mathiasbynens) [reply spec.html:231]: https://github.com/tc39/ecma262/pull/1785

**Review by mathiasbynens — APPROVED** [review]

**Pull request merged**